procrastinate/worker.py

```diff
@@ -284,50 +289,90 @@ async def _fetch_and_process_jobs(self):
-            pass
+        try:
+            # shield the loop task from cancellation
+            # instead, a stop signal is set to enable graceful shutdown
+            await asyncio.shield(loop_task)
+        except asyncio.CancelledError:
+            self.stop()
+            await loop_task
+            raise
+
+    async def _shutdown(self, side_tasks: list[asyncio.Task]):
+        """
+        Gracefully shutdown the worker by cancelling side tasks
+        and waiting for all pending jobs.
+        """
+        await utils.cancel_and_capture_errors(side_tasks)
 
-    async def run(self):
-        self._run_task = asyncio.current_task()
+        now = time.time()
+        for context in self._running_jobs.values():
+            self.logger.info(
+                "Waiting for job to finish: "
+                + context.job_description(current_timestamp=now),
+                extra=self._log_extra(context=None, action="ending_job"),
+            )
 
+        # wait for any in progress job to complete processing
+        # use return_exceptions to not cancel other job tasks if one was to fail
+        await asyncio.gather(
+            *(task for task in self._running_jobs.keys()), return_exceptions=True
+        )
         self.logger.info(
-            f"Starting worker on {utils.queues_display(self.queues)}",
+            f"Stopped worker on {utils.queues_display(self.queues)}",
             extra=self._log_extra(
-                action="start_worker", context=None, queues=self.queues
+                action="stop_worker", queues=self.queues, context=None
             ),
         )
 
-        self.running_jobs = {}
-        self.job_semaphore = asyncio.Semaphore(self.concurrency)
+    async def _start_side_tasks(self) -> list[asyncio.Task]:
+        """Start side tasks such as periodic deferrer and notification listener"""
         side_tasks = [asyncio.create_task(self.periodic_deferrer())]
         if self.wait and self.listen_notify:
             listener_coro = self.app.job_manager.listen_for_jobs(
-                event=self.notify_event,
+                event=self._notify_event,
                 queues=self.queues,
             )
             side_tasks.append(asyncio.create_task(listener_coro, name="listener"))
+        return side_tasks
 
-        context = contextlib.nullcontext()
-        if self.install_signal_handlers:
-            context = signals.on_stop(self.stop)
+    async def _run_loop(self):
+        """
+        Run all side coroutines, then start fetching/processing jobs in a loop
+        """
+        self.logger.info(
+            f"Starting worker on {utils.queues_display(self.queues)}",
+            extra=self._log_extra(
+                action="start_worker", context=None, queues=self.queues
+            ),
+        )
+        self._notify_event.clear()
+        self._stop_event.clear()
+        self._running_jobs = {}
+        self._job_semaphore = asyncio.Semaphore(self.concurrency)
```

Comment:
ha we already recreate the semaphore on starting the worker. this partly answers my previous comment.